### PR TITLE
add JSON communication between server and javascript

### PR DESCRIPTION
**Pull request by pranjali-sharma**
Server returns hardcoded data as JSON, and js file fetches this data and adds it to HTML

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 package com.google.sps.servlets;
 
+import java.util.*;
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:17]: It is better to import only the class you need. Use,
#import java.util.ArrayList.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -23,12 +25,30 @@
...
 public class DataServlet extends HttpServlet {
+   
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String quote="Hello Pranjali!";
-   
-    response.setContentType("text/html;");
+    //Create ArrayList<String> object
+    ArrayList<String> list= new ArrayList<String>();
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: 1) Change this to List<String> list = .... List is the interface and ArrayList is the implementation. It is better to use the interface to represent vars. You only need the methods exposed by the interface.
2) Use a more descriptive variable name. Maybe quotes?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -23,12 +25,30 @@
...
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+   
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String quote="Hello Pranjali!";
-   
-    response.setContentType("text/html;");
+    //Create ArrayList<String> object
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:32]: Remove this comment. It is clear from the code that you are creating an ArrayList. Add comments only to explain things that cannot be easily understood from the code.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -23,12 +25,30 @@
...
+    ArrayList<String> list= new ArrayList<String>();   
+    list.add("And I knew exactly what to do. But in a much more real sense, I had no idea what to do.- Michael G. Scott");
+    list.add("It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.- Sherlock Holmes");
+    list.add("When you are backed against the wall, break the goddamn thing down.- Harvey Specter");
+    list.add("Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.- Tyrion Lannister");
+    //Convert object to JSON
+    String quote=convertToJsonUsingGson(list);
+    System.out.println(quote);
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: Remove the println.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -23,12 +25,30 @@
...
-   
-    response.setContentType("text/html;");
+    //Create ArrayList<String> object
+    ArrayList<String> list= new ArrayList<String>();   
+    list.add("And I knew exactly what to do. But in a much more real sense, I had no idea what to do.- Michael G. Scott");
+    list.add("It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.- Sherlock Holmes");
+    list.add("When you are backed against the wall, break the goddamn thing down.- Harvey Specter");
+    list.add("Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.- Tyrion Lannister");
+    //Convert object to JSON
+    String quote=convertToJsonUsingGson(list);
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:39]: Add space before and after the '='. Run the linter I shared with you and that should take care of formatting issues like this.

Maybe name the variable quotesAsGson or something like that or remove the variable completely and just inline it below like this:
response.getWriter().println(convertToJsonUsingGson(...))

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -23,12 +25,30 @@
...
-    String quote="Hello Pranjali!";
-   
-    response.setContentType("text/html;");
+    //Create ArrayList<String> object
+    ArrayList<String> list= new ArrayList<String>();   
+    list.add("And I knew exactly what to do. But in a much more real sense, I had no idea what to do.- Michael G. Scott");
+    list.add("It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.- Sherlock Holmes");
+    list.add("When you are backed against the wall, break the goddamn thing down.- Harvey Specter");
+    list.add("Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.- Tyrion Lannister");
+    //Convert object to JSON
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: Delete the comment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -23,12 +25,30 @@
...
+    String quote=convertToJsonUsingGson(list);
+    System.out.println(quote);
+    response.setContentType("application/json;");
     response.getWriter().println(quote);
   }
+
+/**
+   * Converts a ServerStats instance into a JSON string using the Gson library. Note: We first added
+   * the Gson library dependency to pom.xml.
+   */
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: I would remove the second part of the comment - "We first added..."
That is part of the docs for using Gson and does not need to be documented here. Think of all the places you might use Gson. You really don't want to add this comment everywhere.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -28,12 +28,26 @@ function addRandomGreeting() {
...
- * Currently just gets a hardcoded message from the server
+ * Fetches random quote from the server.
+ * Currently just gets a hardcoded message from the server.
  * TODO(psharma): Update this method to use Lambdas.
  */
 function getRandomQuote() {
-  fetch('/data').then(response => response.text()).then((quote) => {
-    document.getElementById('quote-container').innerText = quote;
+    console.log("inside getrandomquote");
+  fetch('/data').then(response => response.json()).then((quote) => {
```
Comment (sbahaud) [inline portfolio/src/main/webapp/script.js:37]: Remove logging if you don't need it. I am ok with it if you want to keep it around for the next step in the tutorial.
If you are planning to keep it, make sure the indent is correct. I would suggest running the linter.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -28,12 +28,26 @@ function addRandomGreeting() {
...
+      console.log(quote)
+    const arrayListElement = document.getElementById('quote-container');
+    arrayListElement.innerHTML = '';
```
Comment (sbahaud) [inline portfolio/src/main/webapp/script.js:40]: I would call the variable somethin like quoteContainerElement to be more specific.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (sbahaud) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:17]: I don't the change. Please resolve after the change has been made

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (pranjali-sharma) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: I tried using List<String> list = new ArrayList<String>();
But it wouldn't work. I get "incompatible types: java.util.List<java.lang.String>
 cannot be converted to java.util.ArrayList<java.lang.String>" error

Comment (pranjali-sharma) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: Alright, fixed it

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -23,12 +26,28 @@
...
+    quotes.add(
+        "And I knew exactly what to do. But in a much more real sense, I had no idea what to do.- Michael G. Scott");
+    quotes.add(
+        "It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.- Sherlock Holmes");
+    quotes.add(
+        "When you are backed against the wall, break the goddamn thing down.- Harvey Specter");
+    quotes.add(
+        "Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.- Tyrion Lannister");
+    List<String> quotesList = new ArrayList<String>(quotes);
+    response.setContentType("application/json;");
```
Comment (sbahaud) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: You don't need to make a copy of the quotes. You can pass 'quotes' to convertToJson.
response.getWriter().println(convertTijson(quotes));

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -16,8 +16,12 @@
...
  */
 function addRandomGreeting() {
-  const greetings =
-      ['And I knew exactly what to do. But in a much more real sense, I had no idea what to do.', 'It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.', 'When you’re backed against the wall, break the goddamn thing down.', 'Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.'];
+  const greetings = [
+    'And I knew exactly what to do. But in a much more real sense, I had no idea what to do.',
+    'It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.',
+    'When you’re backed against the wall, break the goddamn thing down.',
+    'Never forget what you are. The rest of the world will not. Wear it like armor, and it can never be used to hurt you.'
+  ];
```
Comment (sbahaud) [inline portfolio/src/main/webapp/script.js:25]: is the greetings still hooked up to the html? If not you can remove this method,

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -28,12 +32,24 @@ function addRandomGreeting() {
...
  * TODO(psharma): Update this method to use Lambdas.
  */
 function getRandomQuote() {
-  fetch('/data').then(response => response.text()).then((quote) => {
-    document.getElementById('quote-container').innerText = quote;
+  console.log('inside getrandomquote');
+  fetch('/data').then(response => response.json()).then((quote) => {
+    console.log(quote)
+    const quoteContainerElement = document.getElementById('quote-container');
+    quoteContainerElement.innerHTML = '';
```
Comment (sbahaud) [inline portfolio/src/main/webapp/script.js:44]: I think you should be able to remove the innerHtml = ''';

**Pull request merged**